appservice/src/deploy/localGitDeploy.ts

```diff
@@ -32,30 +33,63 @@ export async function localGitDeploy(client: SiteClient, fsPath: string): Promis
             await ext.ui.showWarningMessage(message, { modal: true }, deployAnyway, DialogResponses.cancel);
         }
         await verifyNoRunFromPackageSetting(client);
-        // tslint:disable-next-line:no-floating-promises
-        localGit.push(remote, 'HEAD:master');
+
+        // wrap two async actions in a promise because both need to run and finish simaltaneously
+        await new Promise((resolve: () => void, reject: (err: IParsedError) => void): void => {
+            ext.outputChannel.appendLine(formatDeployLog(client, (localize('localGitDeploy', `Deploying Local Git repository to "${client.fullName}"...`))));
+            localGit.push(remote, 'HEAD:master').catch(async (error) => {
+                tokenSource.cancel();
+                // tslint:disable-next-line:no-unsafe-any
+                const parsedError: IParsedError = parseAndRemoveCredentialsFromError(error);
+
+                // if the push fails, check if it's due to diverging history to offer force push
+                if (parsedError.message.indexOf('Updates were rejected because the remote contains work that you do') >= 0) {
+                    await forcePush();
+                    resolve();
+                } else {
+                    reject(parsedError);
+                }
+            });
+
+            waitForDeploymentToComplete(client, kuduClient, commitId, tokenSource.token).then(resolve).catch((error: Error) => {
+                const parsedError: IParsedError = parseAndRemoveCredentialsFromError(error);
```

Comment:
> There are also credentials in that error message, so I've masked those when we parse.

We've been over this before, but I think this needs further discussion. We need to make extra sure there's no credentials in error messages/telemetry and I'm not convinced this handles all cases.

As an aside - why didn't we have this problem before your changes?

Reply:
We did for any error message that wasn't a result from "failed to push" errors.  But I think most error messages weren't actually being output because we were handling it customly.

Reply:
Okay well I would prefer if we had a more broad try/catch encompassing most of `localGitDeploy` and then you only need to call `parseAndRemoveCredentialsFromError` once

Also I would like to see some unit tests for `parseAndRemoveCredentialsFromError`. For example, I think `replace` only replaces the first occurrence and not all occurrences?